File: moatbus/serial.py
## Code to (de)serialize bus messages

from enum import IntEnum
from typing import Optional

from .message import BusMessage
from .crc import CRC16

class S(IntEnum):
    IDLE = 0
    INIT = 1
    LEN = 2
    LEN2 = 3
    DATA = 4
    CRC1 = 5
    CRC2 = 6
    DONE = 7

class ERR(IntEnum):
    OVERFLOW = 1
    LOST = 2
    SPURIOUS = 3
    CRC = 4

class SerBus:
    """
    This is a Sans-IO class that implements interfacing a message to a
    serial line instead of the MoaT bus.

    You need to override these methods:

    * set_timeout(Flag)  -- periodically call .timeout if Flag is True / stop if false
    * process(msg)       -- this message has arrived, handle it!
    * process_ack()      -- a message has been transmitted
    * report_error(typ)  -- a problem has ocurred
    * data_out (bytes)   -- send these data

    External code may call these methods:
    * send (msg, prio?)  -- send a message with priority (or not).
    * send_ack ()        -- ack a message
    * char_in (bits)     -- received this character from serial/pipe
    * timeout()          -- when the timer triggers
    """

    def __init__(self):
        # incoming
        self.m_in = None # bus message
        self.crc_in = 0
        self.len_in = 0

        # outgoing
        self.crc_out = 0

        self.s_in = S.IDLE

        self.idle = 0  # counter, to drop partial messages
        self.ack_out = 0  # counter, to send ACKs
        self.alloc_in()

    def report_error(self, typ, **kw):
        """
        OVERRIDE: There's been a comm problem.
        """
        raise RuntimeError("Override me")

    def set_timeout(self, flag):
        """
        OVERRIDE: Arrange to periodically call .timeout, or not,
        depending on whether @fag is True, or not.
        """
        raise RuntimeError("Override me")
    
    def process(self, msg):
        """
        OVERRIDE: Process this message.
        """
        raise RuntimeError("Override me")

    def process_ack(self):
        """
        OVERRIDE: Process this incoming ACK.
        """
        raise RuntimeError("Override me")

    def alloc_in(self):
        self.m_in = BusMessage()
        self.crc_in = CRC16()
        self.m_in.start_add()

    def send(self, msg, prio):
        """
        Queue a message
        """
        self.data_out(self.send_data(msg, prio))

    def send_ack(self):
        """
        Queue an ACK
        """
        self.data_out(b'\x06')


    def char_in(self, ci:int):
        """
        process an incoming serial character
        """
        self.idle = 0

        if self.s_in == S.IDLE:
            if ci == 6:
                self.process_ack()
            elif ci > 0 and ci <= 0x04:
                self.s_in = S.LEN
            else:
                self.report_error(ERR.SPURIOUS)

        elif self.s_in == S.LEN:
            self.set_timeout(True)

            if ci & 0x80:
                self.len_in = (ci & 0x7F) << 8
                self.s_in = S.LEN2
            else:
                self.len_in = ci
                self.s_in = S.DATA

        elif self.s_in == S.LEN2:
            self.len_in |= ci
            self.s_in = S.DATA

        elif self.s_in == S.DATA:
            self.m_in.add_chunk(ci,8)
            self.crc_in.update(ci)
            self.len_in -= 1
            if self.len_in == 0:
                self.s_in = S.CRC1
                self.crc_in = self.crc_in.finish()

        elif self.s_in == S.CRC1:
            self.crc_in ^= ci<<8
            self.s_in = S.CRC2

        elif self.s_in == S.CRC2:
            self.crc_in ^= ci
            self.set_timeout(False)

            if self.crc_in:
                self.report_error(ERR.CRC)
                self.s_in = S.IDLE
            else:
                self.s_in = S.DONE
                self.process(self.m_in)
            self.alloc_in()
            self.s_in = S.IDLE

        elif self.s_in == S.DONE:
            # ugh, overflow?
            self.report_error(ERR.OVERFLOW)


    def send_data(self, msg, prio) -> bytes:
        """
        Generate chunk of bytes to send for this message.
        """
        prio_data = b"\x01\x02\x03\x04"

        res = bytearray()
        res.append(prio_data[prio])
        n_b = len(msg.data) + msg.header_len
        if n_b >= 0x80:
            res.append(0x80 | (n_b>>8))
            res.append(n_b & 0xFF)
        else:
            res.append(n_b)

        crc = CRC16()
        h = msg.header.bytes
        for b in h:
            crc.update(b)
            res.append(b)

        d = msg.data
        for b in d:
            crc.update(b)
            res.append(b)

        crc = crc.finish()
        res.append(crc>>8)
        res.append(crc&0xFF)
        return res


    def recv(self, prio=0):
        """
        Did we receive a message? if so, return it.
        """
        if self.s_in != S.DONE:
            return None

        msg = self.m_in
        self.alloc_in()
        return msg
        
    def timeout(self):
        """
        Call this periodically (e.g. every 10ms on 9600 baud) whenever
        `set_timeout` told you to.
        """
        if self.s_in != S.IDLE:
            self.idle += 1
            if self.idle > 3:
                self.idle = 0
                self.report_error(ERR.LOST)
                self.s_in = S.IDLE
                self.crc_in = CRC16()
                self.m_in.start_add()
                self.set_timeout(False)
        else:
            self.idle = 0

File: moatbus/serial.c

/*
Serial handler for MoatBus
*/

#include <sys/types.h>
#include <stddef.h>
#include <stdlib.h>
#include <assert.h>

#include "moatbus/type.h"
#include "moatbus/serial.h"
#include "moatbus/crc.h"


static void sb_clear_in(SerBus sb)
{
    sb->crc_in = 0;
    sb->s_in = S_IDLE;
    msg_start_add(sb->m_in);
}

static void sb_alloc_in(SerBus sb)
{
    BusMessage msg = msg_alloc(20);
    if (sb->m_in_first != NULL)
        sb->m_in->next = msg;
    else
        sb->m_in_first = msg;
    sb->m_in = msg;
    sb_clear_in(sb);
}

// Queue+send a message
SerBus sb_alloc(void)
{
    SerBus sb = calloc(sizeof(*sb),1);
    sb_alloc_in(sb);
    return sb;
}

void sb_free(SerBus sb)
{
    while (sb->m_in_first) {
        BusMessage msg = sb->m_in_first;
        sb->m_in_first = msg->next;
        msg_free(msg);
    }
    while (sb->m_out) {
        BusMessage msg = sb->m_out;
        sb->m_out = msg->next;
        msg_free(msg);
    }
    free(sb);
}

// start sending this frame
void sb_send(SerBus sb, BusMessage msg)
{
    if (sb->m_out != NULL)
        sb->m_out_last->next = msg;
    else
        sb->m_out = msg;
    sb->m_out_last = msg;
    msg->next = NULL;

    if (sb->s_out == S_IDLE)
        sb->s_out = S_INIT;
}

// process an incoming serial character
void sb_byte_in(SerBus sb, u_int8_t c)
{
    sb->idle = 0;
    switch(sb->s_in) {
    case S_IDLE:
        if (c == 0x06)
            sb->ack_in++;
        else if (c > 0 && c <= 0x04)
            sb->s_in = S_LEN;
        else
            sb->err_spurious++;
        break;
    case S_INIT:
        sb->s_in = S_LEN;
        break;
    case S_LEN:
        if (c & 0x80) {
            sb->len_in = (c & 0x7F) << 8;
            sb->s_in = S_LEN2;
        } else {
            sb->len_in = c;
            sb->s_in = S_DATA;
        }
        break;
    case S_LEN2:
        sb->len_in |= c;
        sb->s_in = S_DATA;
        break;
    case S_DATA:
        msg_add_chunk(sb->m_in,c,8);
        sb->crc_in = crc16_update(sb->crc_in,c);
        if(!--sb->len_in)
            sb->s_in = S_CRC1;
        break;
    case S_CRC1:
        sb->crc_in ^= c<<8;
        sb->s_in = S_CRC2;
        break;
    case S_CRC2:
        sb->crc_in ^= c;
        if (sb->crc_in)
            sb->err_crc++;
        else
            sb_alloc_in(sb);
        sb_clear_in(sb);
        break;

    case S_DONE: // should not happen
    case S_ACK: // should not happen
        break;
    }
}

// Send char?
int16_t sb_byte_out(SerBus sb)
{
    u_int8_t c;
    u_int16_t len;
    switch(sb->s_out) {
    case S_IDLE:
        assert(sb->m_out == NULL);
        return -1;
    case S_ACK:
        assert (sb->ack_out);
        if(!--sb->ack_out)
            sb->s_out = sb->m_out ? S_INIT : S_IDLE;
        return 0x06;
    case S_INIT:
        assert (sb->m_out != NULL);
        msg_start_extract(sb->m_out);
        sb->s_out = S_LEN;
        sb->crc_out = 0;
        return (sb->m_out->prio & 0x03) +1;
    case S_LEN:
        len = msg_bits(sb->m_out) >> 3;
        if (len >= 0x80) {
            sb->s_out = S_LEN2;
            return 0x80 | (len>>7);
        } else {
            sb->s_out = S_DATA;
            return len;
        }
    case S_LEN2:
        len = msg_bits(sb->m_out) >> 3;
        sb->s_out = S_DATA;
        return len & 0xFF;
    case S_DATA:
        c = msg_extract_chunk(sb->m_out,8);
        sb->crc_out = crc16_update(sb->crc_out, c);
        if (!msg_extract_more(sb->m_out))
            sb->s_out = S_CRC1;
        return c;
    case S_CRC1:
        sb->s_out = S_CRC2;
        return sb->crc_out >> 8;
    case S_CRC2:
        c = sb->crc_out & 0xFF;

        BusMessage msg = sb->m_out;
        sb->m_out = msg->next;
        msg_free(msg);
        if (sb->ack_out)
            sb->s_out = S_ACK;
        else if (sb->m_out == NULL)
            sb->s_out = S_IDLE;
        else
            sb->s_out = S_INIT;
        return c;

    case S_DONE:  // should not happen
        break;
    }
    return 0x00;  // should not happen
}

// Received message?
BusMessage sb_recv(SerBus sb)
{
    if (sb->m_in == sb->m_in_first)
        return NULL;

    BusMessage msg = sb->m_in_first;
    sb->m_in_first = msg->next;
    sb->ack_out ++;
    if (sb->s_out == S_IDLE)
        sb->s_out = S_ACK;
    
    msg_read_header(msg);
    return msg;
}

u_int8_t sb_recv_ack(SerBus sb)
{
    u_int8_t c = sb->ack_in;
    sb->ack_in = 0;
    return c;
}

bool sb_idle(SerBus sb) {
    if (sb->s_in != S_IDLE) {
        if(++sb->idle > 3) {
            sb->idle = 0;
            sb->err_lost++;
            sb->s_in = S_IDLE;
            sb->crc_in = 0;
            msg_start_add(sb->m_in);
        }
        return true;
    } else {
        sb->idle = 0;
        return false;
    }
}

